python/ray/autoscaler/aws/node_provider.py

```diff
@@ -40,15 +44,55 @@ def __init__(self, provider_config, cluster_name):
         # Try availability zones round-robin, starting from random offset
         self.subnet_idx = random.randint(0, 100)
 
+        self.tag_cache = {}  # Tags that we believe to actually be on EC2.
+        self.tag_cache_pending = {}  # Tags that we will soon upload.
+        self.tag_cache_lock = threading.Lock()
+        self.tag_cache_update_event = threading.Event()
+        self.tag_cache_kill_event = threading.Event()
+        self.tag_update_thread = threading.Thread(
+            target=self._node_tag_update_loop
+        )
+        self.tag_update_thread.start()  # TODO: monitor this?
+
         # Cache of node objects from the last nodes() call. This avoids
         # excessive DescribeInstances requests.
         self.cached_nodes = {}
 
-        # Cache of ip lookups. We assume IPs never change once assigned.
-        self.internal_ip_cache = {}
-        self.external_ip_cache = {}
+    def _node_tag_update_loop(self):
+        while True:
+            self.tag_cache_update_event.wait()
+            self.tag_cache_update_event.clear()
+
+            DD = defaultdict(list)
+
+            with self.tag_cache_lock:
+                if self.tag_cache_pending:
+                    for node_id, tags in self.tag_cache_pending.items():
+                        for x in tags.items():
+                            DD[x].append(node_id)
+
+                    for node_id, tags in self.tag_cache_pending.items():
+                        self.tag_cache[node_id].update(tags)
+
+                    self.tag_cache_pending = {}
+
+            for (k, v), node_ids in DD.items():
+                m = "Set tag {}={} on {}".format(k, v, node_ids)
+                with LogTimer("AWSNodeProvider", m) as _:
+                    if k == TAG_RAY_NODE_NAME:  # TODO: to_aws_format
```

Comment:
We should do this TODO right?

Reply:
It's actually done in the line immediately below.

I don't think that this conversion Name <-> TAG_RAY_NODE_NAME is frequent/complex enough that it needs to be factored out. We can either delete the comment or make another conversion function. No preference.